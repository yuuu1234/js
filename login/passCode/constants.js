export const PASSCODE_MAX_LENGTH = 6;

export default {
  PASSCODE_MAX_LENGTH,
};
