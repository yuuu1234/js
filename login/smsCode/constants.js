import numeral from 'numeral';

export const COUNTDOWN_TEXT = 'Code Expires in';
export const UNITTEXT = 'secs';
export const SECONDS_LIMIT = 60;
export const NUMBER_OF_DIGITS = 6;
export const TITLE_USER_SMS_INPUT_SCREEN = 'We Are Sending You An SMS Code';
export const FIREBASE_ERROR_MESSAGES = {
  'auth/invalid-phone-number': 'The phone number is too long.',
  'auth/invalid-verification-code': 'The SMS code is invalid. Please double check or resend it.',
  default: 'There has been an error processing your request.',
};

export function isSmsCodeValid(code, length) {
  return code && code.length === length && code === numeral(code).value().toString();
}
