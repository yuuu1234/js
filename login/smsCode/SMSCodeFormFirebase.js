import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { observer, inject } from 'mobx-react';
import {
  Alert, KeyboardAvoidingView, Text, TouchableOpacity, View,
} from 'react-native';
import { prefixBearerToken } from '@src/utils/helper';
import * as SecureStore from 'expo-secure-store';
import styles from '@modules/login/authTemplate/style';
import { FormButton } from '@src/components';
import CountDownTimer from '@src/components/countDownTimer';
import { PasscodeLogin } from '@src/constants/navigatorConstants';
import {
  FIREBASE_ERROR_MESSAGES,
  COUNTDOWN_TEXT,
  isSmsCodeValid,
  NUMBER_OF_DIGITS,
  SECONDS_LIMIT,
  UNITTEXT,
} from '@src/modules/login/smsCode/constants';

import { TOKEN } from '@src/constants';
import firebase from '../../../constants/firebase';
import SMSCodeInput from '../components/smsCodeInput/basicSMSCodeInput';

const SMSCodeFormFirebase = ({ navigation, loginStore: { login, saveInputChange } }) => {
  const [smsCode, setSmsCode] = useState('');
  const [isTimeUp, setIsTimeUp] = useState(false);

  function onResendSmsCodeClick() {

  }

  async function verifySmsCode() {
    if (login.verifyId && isSmsCodeValid(smsCode, NUMBER_OF_DIGITS)) {
      try {
        const credential = firebase.auth.PhoneAuthProvider.credential(login.verifyId, smsCode);
        const res = await firebase.auth().signInWithCredential(credential);
        if (res.user) {
          const accessToken = await res.user.getIdToken();
          if (accessToken) {
            await SecureStore.setItemAsync(TOKEN, prefixBearerToken(accessToken));
            saveInputChange({ verifyId: null });
            navigation.navigate(PasscodeLogin); // conveniently switching to next screen
          }
        }
      } catch (e) {
        console.log({ e });
        Alert.alert('Error', FIREBASE_ERROR_MESSAGES[e.code] || FIREBASE_ERROR_MESSAGES.default);
      }
    }
  }

  const countdownTimer = (
    <View style={styles.resendCodeContainer}>
      {isTimeUp ? (
        <TouchableOpacity
          testID="resend-code"
          onPress={onResendSmsCodeClick}
          style={styles.resendContainer}
          accessibilityLabel="btn-resendSMSCode"
        >
          <Text style={styles.resendButton}>Resend Code</Text>
        </TouchableOpacity>
      ) : (
        <CountDownTimer
          title={COUNTDOWN_TEXT}
          unitText={UNITTEXT}
          secondsLimit={SECONDS_LIMIT}
          onTimesUp={() => setIsTimeUp(true)}
        />
      )}
    </View>
  );

  const isCodeValid = isSmsCodeValid(smsCode, NUMBER_OF_DIGITS);

  return (
    <KeyboardAvoidingView behavior="padding" enabled style={styles.container}>
      <SMSCodeInput
        onInputChange={(code) => setSmsCode(code)}
        numberDigits={NUMBER_OF_DIGITS}
      />
      <View>
        {countdownTimer}
        <FormButton
          title="Verify"
          disable={!isCodeValid}
          onPress={verifySmsCode}
        />
      </View>
    </KeyboardAvoidingView>
  );
};

SMSCodeFormFirebase.propTypes = {
  navigation: PropTypes.shape({
    navigate: PropTypes.func,
  }).isRequired,
  loginStore: PropTypes.shape({
    login: PropTypes.object.isRequired,
    saveInputChange: PropTypes.func.isRequired,
  }).isRequired,
};

export default inject('loginStore')(observer(SMSCodeFormFirebase));
