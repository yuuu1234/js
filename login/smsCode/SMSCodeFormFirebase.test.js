import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import * as navigatorConstants from '@src/constants/navigatorConstants';
import { Alert } from 'react-native';
import * as firebase from 'firebase';
import * as SecureStore from 'expo-secure-store';
import { act } from 'react-dom/test-utils';
import SMSCodeFormFirebase from './SMSCodeFormFirebase';

jest.spyOn(Alert, 'alert');
describe('SMSCodeFormFirebase', () => {
  const props = {
    navigation: {
      navigate: jest.fn(),
    },
    loginStore: {
      login: { verifyId: 123 },
      saveInputChange: jest.fn(),
    },
    accountStore: {
      defaultAccountDetail: {},
    },
  };
  const component = <SMSCodeFormFirebase {...props} />;
  let wrapper;

  beforeEach(() => {
    wrapper = mount(component);
  });

  it('should render component correctly', () => {
    const tree = renderer.create(component).toJSON();

    expect(tree).toMatchSnapshot();
  });

  describe('isTimeUp', () => {
    it('should render CountDownTimer by default', () => {
      const countDownTimer = wrapper.find('CountdownTimer');
      const resendCode = wrapper.find({ testID: 'resend-code' });

      expect(countDownTimer.exists()).toBeTruthy();
      expect(resendCode.exists()).toBeFalsy();
    });

    it('should render resendCode if time is up', () => {
      let countDownTimer = wrapper.find('CountdownTimer');
      let resendCode = wrapper.find({ testID: 'resend-code' });

      expect(countDownTimer.exists()).toBeTruthy();
      expect(resendCode.exists()).toBeFalsy();

      countDownTimer.props().onTimesUp();
      wrapper.update();
      countDownTimer = wrapper.find('CountdownTimer');
      resendCode = wrapper.find({ testID: 'resend-code' });

      expect(countDownTimer.exists()).toBeFalsy();
      expect(resendCode.exists()).toBeTruthy();
    });

    it('should call onResendSmsCodeClick if user click resend', () => {
      let countDownTimer = wrapper.find('CountdownTimer');
      let resendCode = wrapper.find({ testID: 'resend-code' });
      countDownTimer.props().onTimesUp();
      wrapper.update();
      countDownTimer = wrapper.find('CountdownTimer');
      resendCode = wrapper.find({ testID: 'resend-code' }).at(0);
      resendCode.simulate('click');
      expect(resendCode.exists()).toBeTruthy();
    });
  });

  describe('verifySmsCode', () => {
    firebase.auth.PhoneAuthProvider.credential = jest.fn().mockReturnValue('credential');
    firebase.auth().signInWithCredential = jest.fn().mockReturnValue({
      user: {
        getIdToken: jest.fn().mockReturnValue('idToken'),
      },
    });
    SecureStore.setItemAsync = jest.fn();
    it('should alert error message if something went wrong during sms validation', async () => {
      firebase.auth().signInWithCredential = jest.fn()
        .mockReturnValue(Promise.reject(new Error({ code: 1 })));
      const smsInput = wrapper.find('SMSCodeInput');
      smsInput.props().onInputChange('123456');
      const verifyButton = wrapper.find('FormButton');
      await verifyButton.simulate('click');

      expect(Alert.alert).toHaveBeenCalled();
    });

    it('should do nothing if smsCode not valid', async () => {
      const smsInput = wrapper.find('SMSCodeInput');
      smsInput.props().onInputChange('123');
      const verifyButton = wrapper.find('FormButton');
      await verifyButton.simulate('click');

      expect(Alert.alert).not.toHaveBeenCalled();
      expect(props.navigation.navigate).not.toHaveBeenCalled();
    });

    it('should do nothing if response not contain user', async () => {
      firebase.auth().signInWithCredential = jest.fn().mockReturnValue({});
      const smsInput = wrapper.find('SMSCodeInput');
      smsInput.props().onInputChange('123456');
      const verifyButton = wrapper.find('FormButton');
      await verifyButton.simulate('click');

      expect(Alert.alert).not.toHaveBeenCalled();
      expect(props.navigation.navigate).not.toHaveBeenCalled();
      expect(firebase.auth().signInWithCredential).toHaveBeenCalled();
      expect(firebase.auth.PhoneAuthProvider.credential).toHaveBeenCalled();
    });

    it('should do nothing if can not get access token', async () => {
      firebase.auth().signInWithCredential = jest.fn().mockReturnValue({
        user: {
          getIdToken: jest.fn().mockReturnValue(''),
        },
      });
      const smsInput = wrapper.find('SMSCodeInput');
      smsInput.props().onInputChange('123456');
      const verifyButton = wrapper.find('FormButton');
      await verifyButton.simulate('click');

      expect(Alert.alert).not.toHaveBeenCalled();
      expect(props.navigation.navigate).not.toHaveBeenCalled();
      expect(SecureStore.setItemAsync).not.toHaveBeenCalled();
      expect(firebase.auth().signInWithCredential).toHaveBeenCalled();
      expect(firebase.auth.PhoneAuthProvider.credential).toHaveBeenCalled();
    });

    it('should navigate to passCodeLogin page if all works fine', async () => {
      firebase.auth.PhoneAuthProvider.credential = jest.fn().mockReturnValue('credential');
      firebase.auth().signInWithCredential = jest.fn().mockReturnValue({
        user: {
          getIdToken: jest.fn().mockReturnValue('token'),
        },
      });
      const smsInput = wrapper.find('SMSCodeInput');
      smsInput.props().onInputChange('123456');
      const verifyButton = wrapper.find('FormButton');
      await act(async () => {
        await verifyButton.simulate('click');
      });

      expect(Alert.alert).not.toHaveBeenCalled();
      expect(SecureStore.setItemAsync).toHaveBeenCalled();
      expect(props.navigation.navigate).toHaveBeenCalled();
      expect(firebase.auth().signInWithCredential).toHaveBeenCalled();
      expect(firebase.auth.PhoneAuthProvider.credential).toHaveBeenCalled();
    });
  });
});
