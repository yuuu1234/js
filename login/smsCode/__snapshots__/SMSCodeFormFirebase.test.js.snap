// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SMSCodeFormFirebase should render component correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View>
    <Text
      style={
        Object {
          "alignSelf": "center",
          "color": "#003D9A",
          "fontSize": 22,
          "fontWeight": "bold",
          "marginBottom": 54,
          "maxWidth": 277,
          "textAlign": "center",
        }
      }
    >
      We Are Sending You An SMS Code
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <View
        accessibilityLabel="btn-inputSMSCode-0"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#F8F8F8",
            "borderColor": "#C4C4C4",
            "borderRadius": 5,
            "borderWidth": 1,
            "display": "flex",
            "height": 50,
            "justifyContent": "center",
            "margin": 5,
            "opacity": 1,
            "width": 50,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 24,
              "fontWeight": "bold",
            }
          }
        >
          
        </Text>
      </View>
      <View
        accessibilityLabel="btn-inputSMSCode-1"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#F8F8F8",
            "borderColor": "#C4C4C4",
            "borderRadius": 5,
            "borderWidth": 1,
            "display": "flex",
            "height": 50,
            "justifyContent": "center",
            "margin": 5,
            "opacity": 1,
            "width": 50,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 24,
              "fontWeight": "bold",
            }
          }
        >
          
        </Text>
      </View>
      <View
        accessibilityLabel="btn-inputSMSCode-2"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#F8F8F8",
            "borderColor": "#C4C4C4",
            "borderRadius": 5,
            "borderWidth": 1,
            "display": "flex",
            "height": 50,
            "justifyContent": "center",
            "margin": 5,
            "opacity": 1,
            "width": 50,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 24,
              "fontWeight": "bold",
            }
          }
        >
          
        </Text>
      </View>
      <View
        accessibilityLabel="btn-inputSMSCode-3"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#F8F8F8",
            "borderColor": "#C4C4C4",
            "borderRadius": 5,
            "borderWidth": 1,
            "display": "flex",
            "height": 50,
            "justifyContent": "center",
            "margin": 5,
            "opacity": 1,
            "width": 50,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 24,
              "fontWeight": "bold",
            }
          }
        >
          
        </Text>
      </View>
      <View
        accessibilityLabel="btn-inputSMSCode-4"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#F8F8F8",
            "borderColor": "#C4C4C4",
            "borderRadius": 5,
            "borderWidth": 1,
            "display": "flex",
            "height": 50,
            "justifyContent": "center",
            "margin": 5,
            "opacity": 1,
            "width": 50,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 24,
              "fontWeight": "bold",
            }
          }
        >
          
        </Text>
      </View>
      <View
        accessibilityLabel="btn-inputSMSCode-5"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#F8F8F8",
            "borderColor": "#C4C4C4",
            "borderRadius": 5,
            "borderWidth": 1,
            "display": "flex",
            "height": 50,
            "justifyContent": "center",
            "margin": 5,
            "opacity": 1,
            "width": 50,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 24,
              "fontWeight": "bold",
            }
          }
        >
          
        </Text>
      </View>
      <TextInput
        allowFontScaling={true}
        autoCorrect={false}
        blurOnSubmit={true}
        keyboardType="number-pad"
        maxLength={6}
        onChangeText={[Function]}
        rejectResponderTermination={true}
        style={
          Object {
            "opacity": 0,
            "position": "absolute",
          }
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
  </View>
  <View>
    <View
      style={
        Object {
          "paddingBottom": 50,
          "textAlign": "center",
        }
      }
    >
      <View
        style={
          Object {
            "display": "flex",
            "flexDirection": "row",
            "flexWrap": "wrap",
            "justifyContent": "center",
          }
        }
      >
        <Text>
          Code Expires in 
        </Text>
        <Text
          style={
            Object {
              "color": "#003D9A",
              "fontSize": 14,
              "fontWeight": "bold",
            }
          }
        >
          60 
        </Text>
        <Text
          style={
            Object {
              "color": "#003D9A",
              "fontSize": 14,
              "fontWeight": "bold",
            }
          }
        >
          secs
        </Text>
      </View>
    </View>
    Object {
      "children": Array [
        <Text
          style={
            Object {
              "color": "#fff",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          Verify
        </Text>,
      ],
      "props": Object {
        "accessibilityLabel": "FormButton",
        "accessible": true,
        "focusable": true,
        "onClick": [Function],
        "onResponderGrant": [Function],
        "onResponderMove": [Function],
        "onResponderRelease": [Function],
        "onResponderTerminate": [Function],
        "onResponderTerminationRequest": [Function],
        "onStartShouldSetResponder": [Function],
        "style": Array [
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "backgroundColor": "#003D9A",
            "borderRadius": 15,
            "height": 41,
            "justifyContent": "center",
            "marginBottom": 16,
            "marginHorizontal": 49,
            "marginTop": 6,
          },
          Object {
            "backgroundColor": "#97B5E2",
          },
          Object {},
        ],
      },
      "type": "View",
    }
  </View>
</View>
`;
